packages/legacy/core/App/screens/Connection.tsx

```diff
@@ -121,54 +121,78 @@ const Connection: React.FC<ConnectionProps> = ({ navigation, route }) => {
 
   useEffect(() => {
-    if (!state.notificationRecord || !state.isVisible) {
+    if (!state.isVisible) {
       return
     }
 
-    // connectionless proof request, we don't have connectionless offers.
+    // If we have a connection, but no goal code, we should navigate
+    // to Chat
+    if (connection && !(Object.values(GoalCodes) as [string]).includes(oobRecord?.outOfBandInvitation.goalCode ?? '')) {
+      logger?.info('Connection: Handling connection without goal code, navigate to Chat')
+
+      dispatch({ isVisible: false })
+      navigation.getParent()?.dispatch(
+        CommonActions.reset({
+          index: 1,
+          routes: [{ name: Stacks.TabStack }, { name: Screens.Chat, params: { connectionId: connection.id } }],
+        })
+      )
+
+      return
+    }
+
+    // At this point we should be waiting for a notification
+    // to be processed
+    if (!state.notificationRecord) {
+      return
+    }
+
+    // Connectionless proof request, we don't have connectionless offers.
     if (!connection) {
       dispatch({ isVisible: false })
       navigation.replace(Screens.ProofRequest, { proofId: state.notificationRecord.id })
 
       return
     }
 
-    // connection based proof or offer
-    if (oobRecord?.outOfBandInvitation.goalCode) {
-      const { goalCode } = oobRecord.outOfBandInvitation
+    // At this point, we have connection based proof or offer with
+    // a goal code.
 
-      if (goalCode === GoalCodes.proofRequestVerify || goalCode === GoalCodes.proofRequestVerifyOnce) {
-        logger?.info(`Connection: Handling ${goalCode} goal code, navigate to ProofRequest`)
+    if (!oobRecord) {
+      logger?.error(`Connection: No OOB record where one is expected`)
 
-        dispatch({ isVisible: false })
-        navigation.navigate(Screens.ProofRequest, { proofId: state.notificationRecord.id })
-      } else if (goalCode === GoalCodes.credentialOffer) {
-        logger?.info(`Connection: Handling ${goalCode} goal code, navigate to CredentialOffer`)
+      return
+    }
 
-        dispatch({ isVisible: false })
-        navigation.navigate(Screens.CredentialOffer, { credentialId: state.notificationRecord.id })
-      } else {
-        logger?.info(`Connection: Unable to handle ${goalCode} goal code`)
+    const { goalCode } = oobRecord.outOfBandInvitation
 
-        dispatch({ isVisible: false })
-        navigation.getParent()?.dispatch(
-          CommonActions.reset({
-            index: 1,
-            routes: [{ name: Stacks.TabStack }, { name: Screens.Chat, params: { connectionId: connection.id } }],
-          })
-        )
-      }
-    } else {
-      logger?.info('Connection: Handling connection without goal code, navigate to Chat')
+    if (goalCode === GoalCodes.proofRequestVerify || goalCode === GoalCodes.proofRequestVerifyOnce) {
+      logger?.info(`Connection: Handling ${goalCode} goal code, navigate to ProofRequest`)
 
-      navigation.getParent()?.dispatch(
-        CommonActions.reset({
```

Comment:
Why no longer a reset here? I thought we wanted to avoid back navigation to the connection screen

Reply:
I still use reset, it just moved elsewhere in the code.